### ISSUE TITLE
AccessViolation when reading last byte from pointer to BitmapData

### ISSUE THREAD

**Issue by avendel** [dotnet/runtime#84287]
### Description

When attempting to read the last byte from BitmapData using a pointer, a `System.AccessViolationException` exception is thrown.

### Reproduction Steps

```csharp
using var image = new Bitmap(736, 864, PixelFormat.Format32bppArgb);
var bitmapData = image.LockBits(new Rectangle(0, 0, image.Width, image.Height), ImageLockMode.ReadOnly, PixelFormat.Format32bppArgb);
unsafe {
    if (bitmapData.Stride < 0)
        throw new Exception("Negative stride is not supported.");
    byte* start = (byte*)bitmapData.Scan0.ToPointer();
    byte* pointerToLastByte = start + (bitmapData.Stride * bitmapData.Height) - 1;
    byte lastByte = *pointerToLastByte;
}
```

This code crashes at the last line before the closing brace with:

`dotnet run -c Release -f net6.0-windows`

### Expected behavior

No crash. The `lastByte` contains the alpha channel value of the last pixel.

### Actual behavior

Process crashes with `System.AccessViolationException`.

### Regression?

I could reproduce the issue on both .NET 5 and .NET 6, but not on .NET 7 and .NET Core 3.1.

### Configuration

I was able to replicate this issue on multiple operating systems, namely Windows Server 2016, Windows Server 2019, and Windows 10.

### Other information

* The issue can only be reproduced in release builds, not in debug builds.
* The issue can be replicated for certain image sizes such as 712x896 and 736x864, but not for images of size 735x864, for example.
* The IL code is identical for both .NET 7 and .NET 6.
* Adding a `Console.WriteLine(lastByte)` line prevents the program from crashing.
* The disassembly of the JIT-generated instructions is different for builds that crash and those that don't. Specifically, crashes occur with the `mov ecx,dword ptr [rdx+rcx-1]` instruction, whereas there is no crash when using the `movzx ecx,byte ptr [rax+rcx-1]` instruction (note also the difference in dword vs byte).

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib does this sound familiar to anyone?

**Comment by EgorBo** [comment]
Sounds familiar, e.g. https://github.com/dotnet/runtime/pull/64683 (.net 7) or https://github.com/dotnet/runtime/pull/76341 (.net 8 afair)

**Comment by EgorBo** [comment]
Yeah, does sound like https://github.com/dotnet/runtime/pull/64683

**Comment by JulieLeeMSFT** [comment]
@avendel, what is the business impact for you? Can you upgrade to .NET 7 or is there a workaround that works for you?

**Comment by avendel** [comment]
Hello @JulieLeeMSFT, I appreciate your prompt response. Our plan is to upgrade to the next long-term support (LTS) release, which means we will not be upgrading to .NET 7. While I have found a workaround for the particular occurrence of the problem we found, I am concerned that it might surface in other areas.

Can you provide guidance on what kind of code we should be cautious of and if there is a general workaround that can protect against this problem in code similar to what I have used in my example?

**Comment by JulieLeeMSFT** [comment]
@avendel, great that you founda a workaround for this particular case. Since there is a workaround, we will hold to backport unless it significantly impacts your business.

@EgorBo, do you have a guidance on how this kind of AV can be avoided?

**Comment by EgorBo** [comment]
> @EgorBo, do you have a guidance on how this kind of AV can be avoided?

Hard to say, in the given snippet there is a line:
```csharp
    byte lastByte = *pointerToLastByte;
}
```
Since its value is unused JIT can't just remove this load because of a dereference side-effect (possible nullref) so it "promotes " this unused load to be a nullcheck. and nullchecks used to use incorrect instructions. It seems you're 2nd person to complain about this bug since .NET Framework 3.5 that's why we were not too aggressive to backport the fix.

**Comment by avendel** [comment]
> Since its value is unused JIT can't just remove this load because of a dereference side-effect (possible nullref) so it "promotes " this unused load to be a nullcheck.

The minimal repro I attached is a significant simplification of the production code. However, the production code also had an instance where a pointer was dereferenced, but its value was not actually used. Is it accurate to state that this problem can be avoided by utilizing values obtained after dereferencing?

**Comment by EgorBo** [comment]
> is it accurate to state that this problem can be avoided by utilizing values obtained after dereferencing?

Yes, although it depends on other jit optimizations e.g. if jit inlines everything and sees that the actual result is still unused, etc.

**Comment by jakobbotsch** [comment]
@avendel Is this still a problem for you, or have you been able to upgrade to .NET 8/work around the issue?

**Comment by EgorBo** [comment]
Closing then since the fix made it to 8.0
